apm-agent-core/src/test/java/co/elastic/apm/agent/MockTracer.java

```diff
@@ -57,9 +59,25 @@ public static ElasticApmTracer createRealTracer(Reporter reporter) {
      * the configuration.
      */
     public static ElasticApmTracer createRealTracer(Reporter reporter, ConfigurationRegistry config) {
+
+        // use an object pool that does bookkeeping to allow for extra usage checks
+        TestObjectPoolFactory objectPoolFactory = new TestObjectPoolFactory();
+
         return new ElasticApmTracerBuilder()
             .configurationRegistry(config)
             .reporter(reporter)
+            // use testing bookkeeper implementation here so we will check that no forgotten recyclable object
+            // is left behind
+            .withObjectPoolFactory(objectPoolFactory)
+            .withLifecycleListener(LifecycleListener.ClosableAdapter.of(() -> {
+
+                if (reporter instanceof MockReporter) {
+                    ((MockReporter) reporter).assertRecycledAfterDecrementingReferences();
+                }
+
+                // checking proper object pool usage using tracer lifecycle events
+                objectPoolFactory.checkAllPooledObjectsHaveBeenRecycled();
```

Comment:
I think it makes sense to move this into the `if` scope. If the reporter is just a `Mockito.mock(Reporter.class)`, the references are not decremented and `objectPoolFactory.checkAllPooledObjectsHaveBeenRecycled();` would fail.

Reply:
I'm not sure about this one, as there are also some items that are pooled but do not have any references counter (like errors), I think we might change this if we have a case that fails. Right now it seems that only a few places use MockTracer.